### PR TITLE
[v1.27.0] Refactor: IPFS download

### PR DESCRIPTION
**Pull request by Karrenbelt**
## Proposed changes

Refactor and clean up `IPFSTool.download`. 
partially addresses #501, which may be closed once this PR is accepted.

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -116,6 +116,9 @@ jobs:
       - uses: actions/setup-python@v3
         with:
           python-version: "3.10"
+      - uses: actions/setup-go@v3
+        with:
+          go-version: "1.17.7"
```
Comment (Karrenbelt) [inline .github/workflows/workflow.yml:121]: same reason as other repo: [see comment](https://github.com/valory-xyz/open-autonomy/pull/1635#discussion_r1043800061)

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py [hunk 2]
```diff
@@ -303,50 +304,53 @@ def remove_unpinned_files(self) -> None:
...
-
-        if package_path is None:
+        while not target_path.exists() and retries:
+            retries -= 1
+            try:
+                self.client.get(hash_id, target_dir)
+            except ipfshttpclient.exceptions.StatusError as e:
+                logging.error(f"error on download of {hash_id}: {e}")
+        if not target_path.exists():
+            raise DownloadError(f"Failed to download: {hash_id}")
```
Comment (Karrenbelt) [inline plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:333]: retry, raise only if file still does not exist after attempting download #retries amount of times

File: plugins/aea-cli-ipfs/tests/test_aea_cli_ipfs.py [hunk 3]
```diff
@@ -190,15 +191,16 @@ def test_version_did_not_match():
...
     """Test aea ipfs download."""
     runner = CliRunner()
-    with patch("ipfshttpclient.Client.get") as ipfs_get, patch("os.rmdir"), patch(
-        "pathlib.Path.iterdir", return_value=[1]
-    ), patch("shutil.move"), patch(
-        "aea_cli_ipfs.ipfs_utils.IPFSDaemon._check_ipfs", new=lambda *_: None
+    with as_context(
+        patch("shutil.move"),
+        patch("pathlib.Path.iterdir", return_value=[1]),
+        patch("pathlib.Path.exists", side_effect=[False, True, True]),
```
Comment (Karrenbelt) [inline plugins/aea-cli-ipfs/tests/test_aea_cli_ipfs.py:197]: now also have to patch `pathlib.Path.exists`. Removed two unnecessary mocks.

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py [hunk 4]
```diff
@@ -303,50 +304,53 @@ def remove_unpinned_files(self) -> None:
...
-                package_name, *_ = os.listdir(downloaded_path)
-                package_path = str(Path(target_dir) / package_name)
-
-        if package_path is None:
+        while not target_path.exists() and retries:
+            retries -= 1
+            try:
+                self.client.get(hash_id, target_dir)
+            except ipfshttpclient.exceptions.StatusError as e:
+                logging.error(f"error on download of {hash_id}: {e}")
```
Comment (angrybayblade) [inline plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:331]: What if the IPFS download started and failed after downloading some files, will the `target_path` directory removed on failure? if not, than don't you think having `target_path.exists()` in check will cause the loop to stop even with the broken download?

Comment (Karrenbelt) [reply plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:331]: I don't know, this code is pretty nasty. Didn't write test, and I would say it's not sufficiently tested to begin with. Something like CliTest could come in useful. But no rush I suppose, another draft it is.

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py [hunk 5]
```diff
@@ -295,63 +297,64 @@ def remove(self, hash_id: str) -> Dict:
...
-                    shutil.move(str(each_file), target_dir)
-            except shutil.Error as e:  # pragma: nocover
-                if os.path.isdir(downloaded_path):
-                    shutil.rmtree(downloaded_path)
-                raise DownloadError(f"error on move files {str(e)}") from e
-
-        if os.path.isdir(downloaded_path):
-            shutil.rmtree(downloaded_path)
-        return package_path
+        return str(package_path)
```
Comment (Karrenbelt) [inline plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:353]: only God knows what purpose this serves.

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py [hunk 6]
```diff
@@ -295,63 +297,64 @@ def remove(self, hash_id: str) -> Dict:
...
+            except ipfshttpclient.exceptions.StatusError as e:
+                logging.error(f"error on download of {hash_id}: {e}")
+        else:
+            raise DownloadError(f"Failed to download: {hash_id}")
+
+        if target_path.is_dir() and any(target_path.iterdir()):
+            package_name, *_ = target_path.iterdir()
+            package_path = Path(target_dir) / package_name
+        else:
             package_path = target_dir
```
Comment (Karrenbelt) [inline plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:351]: only God knows what purpose this serves.

https://github.com/valory-xyz/open-aea/blob/06094af29028a3906197c9ac2d3950e39f95dce4/plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py#L327-L335

**Review by DavidMinarsch — APPROVED** [review]

**Comment by Karrenbelt** [comment]
This will need to wait till the morning before it's ready. When 489 is in I'll sync this one up.

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py [hunk 7]
```diff
@@ -309,68 +310,55 @@ def remove_unpinned_files(self) -> None:
...
+            # else it is a directory containing a single package path
+            package_path = download_path
+            if fix_path:
+                # assumption is it contains one nested directory: the package
+                paths = list(download_path.glob("*"))
+                assumption_is_valid = len(paths) == 1 and paths[0].is_dir()
+                if not assumption_is_valid:  # pragma: no cover
+                    error_msg = f"Expected a single directory, found: {paths}"
+                    raise DownloadError(error_msg)
+                package_path = paths.pop()
```
Comment (Karrenbelt) [inline plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:343]: added explicit check for the apparent assumption. Relates to this original piece of code
https://github.com/valory-xyz/open-aea/blob/67ac9a1c6545b1a675e078c959059ff330d2805d/plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py#L350-L358

I am of the opinion this should not exist here and be moved to the one place the returned path is actually used, but that is outside of the scope of this PR
https://github.com/valory-xyz/open-aea/blob/beadc9f172e4c331e4e95e73d17bfb4ddbd60d38/aea/cli/fetch.py#L147-L155

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py [hunk 8]
```diff
@@ -309,68 +310,55 @@ def remove_unpinned_files(self) -> None:
...
 
-        if os.path.exists(os.path.join(target_dir, hash_id)):  # pragma: nocover
-            raise DownloadError(f"{hash_id} was already downloaded to {target_dir}")
```
Comment (Karrenbelt) [inline plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:330]: no longer raising here. `ipfshttpclient.Client.get` also overwrites if the path already exists: if your local is different, your local is wrong. 

Furthermore, if we `fix_path` and effectively remove the content up one directory, this check becomes useless.

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py [hunk 9]
```diff
@@ -309,68 +310,55 @@ def remove_unpinned_files(self) -> None:
...
+            package_path.rename(target_dir / package_path.name)
```
Comment (Karrenbelt) [inline plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:345]: renaming a file or directory to one that already exists will throw an error.

File: plugins/aea-cli-ipfs/tests/test_aea_cli_ipfs.py [hunk 10]
```diff
@@ -238,14 +238,6 @@ def new_callable(*_, **__) -> None:
...
         return patch("ipfshttpclient.Client.get", new_callable=lambda: new_callable)
 
-    def test_ipfs_download_target_path_exists(self) -> None:
-        """Test aea ipfs download target_path exists."""
-
-        Path(self.target_dir, self.some_ipfs_hash).mkdir(parents=True)
-        expected = f"{self.some_ipfs_hash} was already downloaded"
-        with pytest.raises(click.ClickException, match=expected):
-            self.run_cli(*self.args, catch_exceptions=False, standalone_mode=False)
-
```
Comment (Karrenbelt) [inline plugins/aea-cli-ipfs/tests/test_aea_cli_ipfs.py:248]: this tests is no longer relevant

File: plugins/aea-cli-ipfs/tests/test_utils.py [hunk 11]
```diff
@@ -228,13 +226,10 @@ def test_tool_download() -> None:
...
+                ipfs_tool.download("some", tmp_dir, attempts=5)
         assert client_mock.get.call_count == 5
 
         client_mock.get = Mock()
 
         ipfs_tool.download("some", tmp_dir, attempts=5)
-
-        with pytest.raises(DownloadError, match="was already downloaded to"):
-            os.mkdir(Path(tmp_dir) / "some")
-            ipfs_tool.download("some", tmp_dir, attempts=5)
```
Comment (Karrenbelt) [inline plugins/aea-cli-ipfs/tests/test_utils.py:240]: and neither is this one

**Comment by Karrenbelt** [comment]
failures
- windows 3.7
    ```
    FAILED tests/test_cli/test_get_multiaddress.py::TestGetMultiAddressCommandConnectionNegative::test_run[fake-password]
    FAILED tests/test_helpers/test_dependency_tree.py::test_generation_of_dependency_tree_of_project
    FAILED tests/test_cli/test_test.py::TestVendorPackageTestByType::test_run[protocol]
    FAILED tests/test_cli/test_test.py::TestVendorPackageTestByType::test_run[connection]
    FAILED tests/test_cli/test_test.py::TestVendorPackageTestByType::test_run[skill]
    FAILED tests/test_cli/test_test.py::TestVendorPackageTestByType::test_run[contract]
    FAILED tests/test_cli/test_test.py::TestPackageTestByType::test_run[protocol]
    FAILED tests/test_cli/test_test.py::TestPackageTestByType::test_run[connection]
    FAILED tests/test_cli/test_test.py::TestPackageTestByType::test_run[skill] - ...
    FAILED tests/test_cli/test_test.py::TestPackageTestByType::test_run[contract]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPathWithCov::test_run[connection]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPathWithCov::test_run[protocol]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPathWithCov::test_run[skill]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPathWithCov::test_run[contract]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPathEmptyTestSuite::test_run[protocol]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPathEmptyTestSuite::test_run[contract]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPathEmptyTestSuite::test_run[skill]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPathEmptyTestSuite::test_run[connection]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPath::test_run[protocol]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPath::test_run[contract]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPath::test_run[connection]
    FAILED tests/test_cli/test_test.py::TestPackageTestByPath::test_run[skill] - ...
    FAILED tests/test_aea_builder.py::TestExtraDeps::test_install_dependency - ae...
    FAILED tests/test_cli/test_generate_all_protocols.py::TestParentAsRootDir::test_root_dir_dont_match
    FAILED tests/test_cli/test_generate_all_protocols.py::TestParentAsRootDir::test_root_dir_parent
    FAILED tests/test_cli/test_generate_all_protocols.py::TestGenerateAllProtcols::test_no_bump
    FAILED tests/test_cli/test_generate_all_protocols.py::TestGenerateAllProtcols::test_check_clean_pass
    FAILED tests/test_cli/test_generate_all_protocols.py::TestGenerateAllProtcols::test_check_bump
    FAILED tests/test_cli/test_generate_all_protocols.py::TestGenerateAllProtcols::test_check_clean_fail
    FAILED tests/test_cli/test_generate_all_protocols.py::TestGenerateAllProtcols::test_no_bump_failure
    FAILED tests/test_cli/test_generate_all_protocols.py::TestGenerateAllProtcols::test_check_bump_fail
    FAILED tests/test_cli/test_generate_all_protocols.py::TestGenerateAllProtcols::test_run
    ERROR tests/test_helpers/test_ipfs/test_base.py::TestDirectoryHashing::test_get_dir_hash[1-1-bafybeies5n7nngqpzwzhety4ndktmzfjgwr5pk2ge3wbftqjwymhlzrc7e]
    ERROR tests/test_helpers/test_ipfs/test_base.py::TestDirectoryHashing::test_depth_0
    ERROR tests/test_helpers/test_ipfs/test_base.py::TestDirectoryHashing::test_depth_multi
    ERROR tests/test_helpers/test_ipfs/test_base.py::TestDirectoryHashing::test_hash_bytes
    ERROR tests/test_helpers/test_ipfs/test_base.py::TestDirectoryHashing::test_get_dir_hash[0-0-QmTGBxU5aqqpeiihQxcWr4xynhqWt23R73Btss2j8r9XcC]
    ERROR tests/test_helpers/test_ipfs/test_base.py::TestDirectoryHashing::test_get_dir_hash[1-0-QmYEAWM6jmjffDQNiyVjVWFcx7SRvutnoRuEP6AJKN9apY]
    ERROR tests/test_helpers/test_ipfs/test_base.py::TestDirectoryHashing::test_depth_1
    ERROR tests/test_helpers/test_ipfs/test_base.py::TestDirectoryHashing::test_get_dir_hash[0-1-bafybeicjexomh6l2rb3efmzojmsx2p2gynjzg3eztf4quu6zyepmnisn4e]
    = 32 failed, 2521 passed, 51 skipped, 51 deselected, 1 warning, 8 errors, 2 rerun in 5512.57s (1:31:52) =
    ```
    one observes `3221225477`, and `hex(3221225477) == "0xc0000005"` (stack overflow)
- windows 3.8 - rare error (see [here](https://github.com/valory-xyz/open-aea/actions/runs/3771843631/jobs/6412487127#step:8:23634))
   ```
               finally:
    >           runner.stop()
    ...
    self = <ProactorEventLoop running=False closed=False debug=False>

        def run_forever(self):
            try:
    >           assert self._self_reading_future is None
    E           AssertionError
    ...
    FAILED tests/test_launcher.py::TestAsyncLauncherMode::test_start_stop
   ```

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py [hunk 12]
```diff
@@ -309,68 +310,55 @@ def remove_unpinned_files(self) -> None:
...
             try:  # download to tmp_dir in case of midway download failure
                 with tempfile.TemporaryDirectory() as tmp_dir:
                     self.client.get(hash_id, tmp_dir)
-                    download_path = Path(tmp_dir) / hash_id
-                    if download_path.is_dir():
-                        shutil.copytree(download_path, downloaded_path)
-                    else:
-                        shutil.copy(download_path, downloaded_path)
-                    break
+                    return move_to_target_dir(Path(tmp_dir) / hash_id)
```
Comment (solarw) [inline plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:356]: suppose `move_to_target_dir(Path(tmp_dir) / hash_id)` should not fail and can be extracted outside attempts loop

if file system operation fails, probably it's not recoverable

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py (hunk 7 above)
Comment (solarw) [reply plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:343]: nice refactor.

but i dont like the mix of functionalities in download method.

i think we need quite simple download with retry method
ans some high level method that uses simple download and applies all the package related things

`download` method should know nothing about packages, files and how handle data downloaded.

**Review by solarw — DISMISSED** [review]
LGTM, couple optional comments

File: plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py (hunk 7 above)
Comment (Karrenbelt) [reply plugins/aea-cli-ipfs/aea_cli_ipfs/ipfs_utils.py:343]: 100% agree (as per #501), it's mostly that I try to limit the amount of changes in a single PR. Now with the nested function it should be easier to achieve in a future refactor.

**Review by DavidMinarsch — DISMISSED** [review]

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**